scripts/linking.py

```diff
@@ -1,86 +1,241 @@
 import json
 import argparse
-from collections import defaultdict
+import os.path
+import datetime
+import numpy as np
+from joblib import dump, load
+from sklearn.feature_extraction.text import TfidfVectorizer
+import nmslib
 from scispacy import data_util
 
-def normalize(text):
-    return " ".join(text.lower().split(" "))
-
-def linking(entity: data_util.MedMentionEntity, umls_concept_dict_by_name):
+def load_umls_kb(umls_path: str):
     """
-    Links the entity mention and returns umls concept
+    Returns a list of concept object
     """
-    normalized_mention = normalize(entity.mention_text)
-    concept_candidates = umls_concept_dict_by_name.get(normalized_mention)
-    return None if concept_candidates is None else concept_candidates[0]
+    with open(umls_path) as f:
+        print(f'Loading umls concepts from {umls_path}')
+        umls_concept_list = json.load(f)
+    print(f'Number of umls concepts: {len(umls_concept_list)}')
+    return umls_concept_list
+
+def nmslis_knn_with_zero_vectors(vectors, k, ann_index):
+    empty_vectors_boolean_flags = np.array(vectors.sum(axis=1) != 0).reshape(-1,)
+    empty_vectors_count = vectors.shape[0] - sum(empty_vectors_boolean_flags)
+    print(f'Number of empty vectors: {empty_vectors_count}')
+
+    # remove empty vectors before calling `ann_index.knnQueryBatch`
+    vectors = vectors[empty_vectors_boolean_flags]
+    neighbors = [x[0].tolist() for x in ann_index.knnQueryBatch(vectors, k=k)]
+
+    # all an empty list in place for each empty vector to make sure len(extended_neighbors) == len(vectors)
 
-def prepare_umls_indices(umls_path: str):
+    # init extended_neighbors with a list of Nones
+    extended_neighbors = np.empty((len(empty_vectors_boolean_flags),), dtype=object)
+
+    # neighbors need to be convected to an np.array of objects instead of ndarray of dimensions len(vectors)xk
+    # Solution: add a row to `neighbors` with any length other than k. This way, calling np.array(neighbors)
+    # returns an np.array of objects
+    neighbors.append([])
+    # interleave `neighbors` and Nones in `extended_neighbors`
+    extended_neighbors[empty_vectors_boolean_flags] = np.array(neighbors)[:-1]
+
+    return extended_neighbors
+
+def generate_candidates(mention_texts, k, tfidf_vectorizer, ann_index, ann_concept_id_list):
+    """Given a list of mention texts, returns a list of candidate neighbors
+    args:
+        mention_texts: list of mention texts
+        k: number of ann neighbors
+        tfidf_vectorizer: text to vector
+        ann_index: approximate nearest neighbor index
+        ann_concept_id_list: a list of concept ids that the ann_index is referencing
     """
-    Returns two indices, one by entity id, and one by concept name.
+    print(f'Generating candidates for {len(mention_texts)} mentions')
+    tfidfs = tfidf_vectorizer.transform(mention_texts)
+    start_time = datetime.datetime.now()
+    neighbors = nmslis_knn_with_zero_vectors(tfidfs, k, ann_index)
```

Comment:
I don't understand how this works (the part where we do knn with zero vectors). Could you explain it a bit?